fixes bugs with share state

bosvensson1 opened the pull request:
Currently Open MW runs shareState on object paging chunks. In theory the SceneManager should have shared this state earlier. Investigations revealed a few issues:
1. The Optimizer modified shared statesets.
2. Because of an API limitation, the Optimizer is unable to optimise the root node of the passed in scene graph. During object paging, we wrap the loaded nodes under a new root node. With this new root node, object paging was able to perform optimisations that we could theoretically perform earlier in the scene manager.
3. Currently, Open MW runs state sharing before optimisation. There are issues with this approach because there is an optimisation stage (the same one which caused bug no. 1) we should ideally run before state sharing.

With these changes all above issues should be resolved. We should test the following before merging this PR:

- Bug 6048 that @cemocfr solved by adding shareState to object paging should not reoccur.
- State Graphs, Nodes and Drawables count in the osg stats is not higher than in master.

AnyOldName3 commented:
This seems like a good idea. However it looks like the kind of change that might introduce surprise regressions, so I'd rather some of our testers played around with this before it's merged.

AbduSharif commented:
I tried vanilla and a somewhat heavily modded setups and both worked fine, I did get slightly better FPS (1-3).

bosvensson1 commented:
Expressing FPS delta without a reference is meaningless. For example, experiencing 1000 FPS versus 1002 FPS would not be statistically significant. On the other hand, experiencing 10 FPS versus 12 FPS would be highly significant.

Could you open the osg stats four times to compare the following from the same vantage point?

> State Graphs, Nodes and Drawables count

AbduSharif commented:
I'm not/wasn't testing for performance but for regressions, Also the profiler is unreadable on Android because of a text bug and how it's scale is broken. If you don't like side notes like how I did get better FPS then I'll stop mentioning it.

bosvensson1 commented:
I fixed another issue related to the order of optimisation routines. Further, I reverted my change that allowed optimising out root nodes. Producing a naked osg::Geometry as a scene template may be counterproductive for the cloning operations currently mandated by the engine. An osg::Group allows shallower copying than an osg::Geometry.

Please test once more. Specifically, we should confirm the following goals:
- Bug 6048 that @cemocfr solved by adding shareState to object paging should not reoccur.
- State Graphs, Nodes and Drawables count in the osg stats is not higher than in master.

bosvensson1 commented:
I realised we should conduct all tests for this PR with `object paging min size merge factor = 1.0`. Otherwise, discrepancies in the number of objects included by object paging may skew the test results.

bosvensson1 commented:
> I realised we should conduct all tests for this PR with object paging min size merge factor = 1.0. 

Without this setting, changes which reduce the count of StateSets will simply make object paging include more objects until the count is roughly as previously, skewing the comparison.

psi29a commented:
So far, no surprises in rendering so that's good.
